### PR TITLE
Fix race condition on early connection failure

### PR DESCRIPTION
**Pull request by pm47**
Both `Client` and `TransportHandler` were watching the connection actor,
which resulted in undeterministic behavior during termination of
`PeerConnection`.

We now always return a message when a connection fails during
authentication.

Took the opportunity to add more typing (insert
deathtoallthestring.jpg).

### CODE REVIEW

File: eclair-core/src/main/scala/fr/acinq/eclair/io/Client.scala [hunk 1]
```diff
@@ -112,24 +115,28 @@ class Client(nodeParams: NodeParams, switchboard: ActorRef, router: ActorRef, re
       SupervisorStrategy.Stop
   }
 
+
+  override def postStop(): Unit = {
```
Comment (t-bast) [inline eclair-core/src/main/scala/fr/acinq/eclair/io/Client.scala:119]: Do you want to keep that?

File: eclair-core/src/main/scala/fr/acinq/eclair/io/Client.scala [hunk 2]
```diff
@@ -75,32 +74,36 @@ class Client(nodeParams: NodeParams, switchboard: ActorRef, router: ActorRef, re
...
             case Tcp.CommandFailed(_: Socks5Connect) =>
               log.info(s"connection failed to ${str(remoteAddress)} via SOCKS5 ${str(proxyAddress)}")
-              origin_opt.foreach(_ ! Status.Failure(ConnectionFailed(remoteAddress)))
+              origin_opt.foreach(_ ! PeerConnection.ConnectionResult.ConnectionFailed(remoteAddress))
               context stop self
             case Socks5Connected(_) =>
               log.info(s"connected to ${str(remoteAddress)} via SOCKS5 proxy ${str(proxyAddress)}")
-              auth(proxy)
-              context become connected(proxy)
+              context unwatch proxy
```
Comment (t-bast) [inline eclair-core/src/main/scala/fr/acinq/eclair/io/Client.scala:81]: What happens to the proxy then? We're now calling `auth` on `connection` instead of `proxy`, is that ok?

File: eclair-core/src/main/scala/fr/acinq/eclair/io/Client.scala (hunk 1 above)
Comment (pm47) [reply eclair-core/src/main/scala/fr/acinq/eclair/io/Client.scala]: Done with 3e6a4b4.

File: eclair-core/src/main/scala/fr/acinq/eclair/io/Client.scala (hunk 2 above)
Comment (pm47) [reply eclair-core/src/main/scala/fr/acinq/eclair/io/Client.scala:81]: Oops, fixed with a113b8c.

We are obviously missing a test, but it is easier said than done. Would need to mock a socks5 proxy?

Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/io/Client.scala:81]: It's not easy, if we simply mock there's a risk our mock deviates from what the real proxy does and our test isn't accurate...it's enough of an edge case that I think we're ok with what we have for now

Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/io/Client.scala:81]: Maybe we should deploy an Acinq node behind Tor, with a small amount of funds, just to make sure we test this E2E often enough?

Comment (pm47) [reply eclair-core/src/main/scala/fr/acinq/eclair/io/Client.scala:81]: We do, both on testnet and mainnet. But our peers can initiate connections too, so this bugs is not easy to detect. It also means it doesn't have a lot of impact.

**Comment by pm47** [comment]
> maybe worth a manual test for Tor if that's not too painful?

will monitor on next deployment
